### PR TITLE
synchronous socket is deterministic test

### PR DESCRIPTION
**Pull request by patduin**
@massdosage reported tests still failing sporadicly, hopefully this really gets rid of it

### CODE REVIEW

File: waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java [hunk 1]
```diff
@@ -45,104 +38,49 @@
...
-        requestCount--;
-      } catch (Exception e) {
-        throw new RuntimeException("Error while waiting for request completion", e);
+      SocketChannel socketChannel = serverSocketChannel.accept();
+      while (socketChannel != null) {
+        requests++;
+        try (InputStream input = Channels.newInputStream(socketChannel)) {
+          ByteStreams.copy(input, output);
+        }
+        socketChannel = serverSocketChannel.accept();
```
Comment (massdosage) [inline waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java]: Why do you need to do this?

File: waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java [hunk 2]
```diff
@@ -45,104 +38,49 @@
...
-      address = (InetSocketAddress) listener.getLocalAddress();
+      serverSocketChannel = (ServerSocketChannel) ServerSocketChannel
+          .open()
+          .bind(new InetSocketAddress(0))
+          .configureBlocking(false);
+      address = (InetSocketAddress) serverSocketChannel.getLocalAddress();
     } catch (IOException e) {
       throw new RuntimeException(e);
     }
   }
```
Comment (massdosage) [inline waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java]: With respect to all the code that's gone, any idea if it served much purpose? Are there new edge cases we might be exposed to? It seemed to be handling cases where you'd want to wait for input that had been previously been sent etc. Not sure we need it all, just curious what you thought after looking into in more detail. This looks much more straightforward and doesn't seem to exhibit the issue with reads returning data out of order that we've been seeing.

File: waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java (hunk 1 above)
Comment (patduin) [reply waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java]: an `accept()` for every request, see the `expectMultipleRequests` test it loops through this 3 times

File: waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java (hunk 2 above)
Comment (patduin) [reply waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java]: hard to tell, there were some bugs on the sending side (where we didn't flush explicitly) so some of them were added to fix that until we fixed the flushing. 
But I am not sure of the whole history. This doesn't seem to block, requests are pending until you `accept` at which point you can read it. It's probably less performant this way but more deterministic which is fine cause we use this is meant for tests.
